File: dd_parser_downscale_table.cpp
#include <cassert>
#include <cstdint>
#include <cstdio>
#include <cmath>
#include <cstring>
#include <vector>
#include <array>
#include <fstream>
#include <string>
#include <inttypes.h>

static inline uint64_t doubleBits(double x) {
	uint64_t u;
	std::memcpy(&u, &x, sizeof(u));
	return u;
}

static inline double bitsToDouble(uint64_t u) {
	double x;
	std::memcpy(&x, &u, sizeof(x));
	return x;
}

static uint64_t roundToEvenFromParts(uint64_t intPart, double fracPart) {
	if (fracPart < 0.5) {
		return intPart;
	}
	if (fracPart > 0.5) {
		return intPart + 1;
	}
	return (intPart & 1) == 0 ? intPart : (intPart + 1);
}

static uint64_t roundShiftRightToEven(uint64_t N, int r) {
	if (r <= 0) {
		return N << (-r);
	}
	uint64_t q = N >> r;
	uint64_t rem = N & ((uint64_t(1) << r) - 1);
	uint64_t half = uint64_t(1) << (r - 1);
	if (rem > half) {
		return q + 1;
	}
	if (rem < half) {
		return q;
	}
	return q + (q & 1);
}

static int ilog2FromDouble(double x) {
	int e2;
	std::frexp(x, &e2);
	return e2 - 1;
}

static inline int pow2ExponentFromBits(double f) {
	int e;
	double m = std::frexp(f, &e);
	assert(m == 0.5);
	return e - 1;
}

struct DoubleDouble {
	double high;
	double low;
	double toDouble() const { return high + low; }
};

static inline double scaleFloat(const DoubleDouble& acc, double factor) {
	return acc.toDouble() * factor;
}

double scaleDDPow2Exact(const DoubleDouble& acc, double factor) {
	assert(factor > 0.0);
	int k = pow2ExponentFromBits(factor);
	uint64_t H = static_cast<uint64_t>(acc.high);
	double L = acc.low;
	if (H == 0 && L == 0.0) {
		return 0.0;
	}
	int e;
	if (H == 0) {
		std::frexp(L, &e);
		e = (e - 1) + k;
	} else {
		e = ilog2FromDouble(acc.high) + k;
	}
	uint64_t N = 0;
	if (e < -1022) {
		int E = k;
		int T = E + 1074;
		double frac;
		if (T >= 0) {
			uint64_t A = H << T;
			double Bf = std::ldexp(L, T);
			uint64_t Bi = static_cast<uint64_t>(Bf);
			frac = Bf - static_cast<double>(Bi);
			N = A + Bi;
		} else {
			int kk = -T;
			uint64_t q = H >> kk;
			uint64_t r = H & ((uint64_t(1) << kk) - 1);
			double Bf = (r + std::ldexp(L, kk)) / static_cast<double>(uint64_t(1) << kk);
			frac = Bf;
			N = q;
		}
		if (frac > 0.5) {
			++N;
		} else if (frac == 0.5 && (N & 1) == 1) {
			++N;
		}
		if (N <= 0) {
			return 0.0;
		}
		if (N >= (uint64_t(1) << 52)) {
			return bitsToDouble(0x0010000000000000ULL);
		}
		return bitsToDouble(N);
	}
	if (H == 0) {
		double mL = std::frexp(L, &e);
		e = (e - 1) + k;
		double t = std::ldexp(mL, 53);
		uint64_t Ni = static_cast<uint64_t>(t);
		double frac = t - static_cast<double>(Ni);
		N = roundToEvenFromParts(Ni, frac);
		if (N == (uint64_t(1) << 53)) {
			N = uint64_t(1) << 52;
			++e;
		}
	} else {
		int exph = ilog2FromDouble(acc.high);
		int s = 52 - exph;
		double frac;
		if (s >= 0) {
			uint64_t A = H << s;
			double Bf = std::ldexp(L, s);
			uint64_t Bi = static_cast<uint64_t>(Bf);
			frac = Bf - static_cast<double>(Bi);
			N = roundToEvenFromParts(A + Bi, frac);
		} else {
			int k2 = -s;
			uint64_t q = H >> k2;
			uint64_t r = H & ((uint64_t(1) << k2) - 1);
			double Bf = (r + std::ldexp(L, k2)) / static_cast<double>(uint64_t(1) << k2);
			N = roundToEvenFromParts(q, Bf);
		}
		e = exph + k;
		if (N == (uint64_t(1) << 53)) {
			N = uint64_t(1) << 52;
			++e;
		}
	}
	int expfield = e + 1023;
	if (expfield <= 0) {
		int rshift = 1 - expfield;
		uint64_t payload = roundShiftRightToEven(N, rshift);
		if (payload <= 0) {
			return 0.0;
		}
		if (payload >= (uint64_t(1) << 52)) {
			return bitsToDouble(0x0010000000000000ULL);
		}
		return bitsToDouble(payload);
	}
	if (expfield >= 0x7ff) {
		return INFINITY;
	}
	uint64_t mant = N - (uint64_t(1) << 52);
	uint64_t bits = (static_cast<uint64_t>(expfield) << 52) | mant;
	return bitsToDouble(bits);
}

struct TestCase {
	uint64_t hbits;
	uint64_t lbits;
	uint64_t fbits;
	uint64_t obits;
};

static std::vector<TestCase> loadTests(const char* path) {
	std::vector<TestCase> tests;
	std::ifstream f(path);
	std::string line;
	while (std::getline(f, line)) {
		TestCase t;
		if (std::sscanf(line.c_str(),
						" [%" SCNx64 ", %" SCNx64 ", %" SCNx64 ", %" SCNx64 "],",
						&t.hbits, &t.lbits, &t.fbits, &t.obits) == 4) {
			tests.push_back(t);
		}
	}
	return tests;
}
int main() {
	const auto tests = loadTests("dd_parser_fuzz_table.txt");
	struct Scaler { const char* name; double (*fn)(const DoubleDouble&, double); };
	const std::array<Scaler, 2> scalers = {{
		{"scale_float", scaleFloat},
		{"scale_dd_pow2_exact", scaleDDPow2Exact},
	}};
	std::array<size_t, scalers.size()> counts = {};
	for (const auto& t : tests) {
		double raw_h = bitsToDouble(t.hbits);
		double raw_l = bitsToDouble(t.lbits);
		double sign = (raw_h < 0.0 || (raw_h == 0.0 && raw_l < 0.0)) ? -1.0 : 1.0;
		DoubleDouble acc(std::fabs(raw_h), std::fabs(raw_l));
		double factor = bitsToDouble(t.fbits);
		for (size_t i = 0; i < scalers.size(); ++i) {
			double y = sign * scalers[i].fn(acc, factor);
			if (doubleBits(y) != t.obits) {
				++counts[i];
			}
		}
	}
	for (size_t i = 0; i < scalers.size(); ++i) {
		std::printf("%s mismatches: %zu\n", scalers[i].name, counts[i]);
	}
	return 0;
}

File: dd_parser_downscale_table.py
#!/usr/bin/env python3
import math, struct
from decimal import Decimal, getcontext

DECIMAL_PREC = 100
getcontext().prec = DECIMAL_PREC

def double_bits(x: float) -> int:
	return struct.unpack(">Q", struct.pack(">d", x))[0]

def bits_to_double(u: int) -> float:
	return struct.unpack(">d", struct.pack(">Q", u))[0]

def round_to_even_from_parts(int_part: int, frac_part: float) -> int:
	if frac_part < 0.5:
		return int_part
	if frac_part > 0.5:
		return int_part + 1
	return int_part if (int_part & 1) == 0 else (int_part + 1)

def round_shift_right_to_even(N: int, r: int) -> int:
	if r <= 0:
		return N << (-r)
	q = N >> r
	rem = N & ((1 << r) - 1)
	half = 1 << (r - 1)
	if rem > half:
		return q + 1
	if rem < half:
		return q
	return q + (q & 1)

def ilog2_from_double(x: float) -> int:
	m, e2 = math.frexp(x)
	return e2 - 1

def pow2_exponent_from_bits(f: float) -> int:
	u = double_bits(f)
	exp = (u >> 52) & 0x7FF
	frac = u & ((1 << 52) - 1)
	if exp == 0:
		assert frac != 0 and (frac & (frac - 1)) == 0, f"factor not pure pow2 (subnormal): bits=0x{u:016x}"
		pos = frac.bit_length() - 1
		return -1074 + pos
	assert frac == 0, f"factor not pure pow2 (normal): bits=0x{u:016x}"
	return exp - 1023

class DoubleDouble:
	def __init__(self, high: float = 0.0, low: float = 0.0):
		self.high = float(high)
		self.low = float(low)

	def __float__(self):
		return self.high + self.low

def scale_float(acc: DoubleDouble, factor: float) -> float:
	return float(acc) * factor

def scale_decimal(acc: DoubleDouble, factor: float) -> float:
	return float((Decimal(acc.high) + Decimal(acc.low)) * Decimal(factor))

def scale_dd_pow2_exact(acc: DoubleDouble, factor: float) -> float:
	assert factor > 0.0
	k = pow2_exponent_from_bits(factor)
	H = int(acc.high)
	L = float(acc.low)
	if H == 0 and L == 0.0:
		return 0.0
	if H == 0:
		mL, eL = math.frexp(L)
		e = (eL - 1) + k
	else:
		e = ilog2_from_double(acc.high) + k
	if e < -1022:
		E = k
		T = E + 1074
		if T >= 0:
			A = H << T
			Bf = math.ldexp(L, T)
			Bi = int(Bf)
			frac = Bf - Bi
			N = A + Bi
		else:
			kk = -T
			q = H >> kk
			r = H & ((1 << kk) - 1)
			Bf = (r + math.ldexp(L, kk)) / float(1 << kk)
			frac = Bf
			N = q
		if frac > 0.5:
			N += 1
		elif frac == 0.5 and (N & 1) == 1:
			N += 1
		if N <= 0:
			return 0.0
		if N >= (1 << 52):
			return bits_to_double(0x0010000000000000)
		return bits_to_double(N)
	if H == 0:
		mL, eL = math.frexp(L)
		e = (eL - 1) + k
		t = math.ldexp(mL, 53)
		Ni = int(t)
		frac = t - Ni
		N = round_to_even_from_parts(Ni, frac)
		if N == (1 << 53):
			N = (1 << 52)
			e += 1
	else:
		exph = ilog2_from_double(acc.high)
		s = 52 - exph
		if s >= 0:
			A = H << s
			Bf = math.ldexp(L, s)
			Bi = int(Bf)
			frac = Bf - Bi
			N = round_to_even_from_parts(A + Bi, frac)
		else:
			k2 = -s
			q = H >> k2
			r = H & ((1 << k2) - 1)
			Bf = (r + math.ldexp(L, k2)) / float(1 << k2)
			N = round_to_even_from_parts(q, Bf)
		e = exph + k
		if N == (1 << 53):
			N = (1 << 52)
			e += 1
	expfield = e + 1023
	if expfield <= 0:
		rshift = 1 - expfield
		payload = round_shift_right_to_even(N, rshift)
		if payload <= 0:
			return 0.0
		if payload >= (1 << 52):
			return bits_to_double(0x0010000000000000)
		return bits_to_double(payload)
	if expfield >= 0x7FF:
		return float('inf')
	mant = N - (1 << 52)
	bits = (expfield << 52) | mant
	return bits_to_double(bits)

def run():
	ns = {}
	with open("dd_parser_fuzz_table.txt") as f:
		exec(f.read(), ns)
	tests = ns["mismatches"] + ns["matches"]
	scalers = [
		("scale_float", scale_float),
		("scale_decimal", scale_decimal),
		("scale_dd_pow2_exact", scale_dd_pow2_exact),
	]
	counts = {name: 0 for name, _ in scalers}
	for hbits, lbits, fbits, obits in tests:
		raw_h = bits_to_double(hbits)
		raw_l = bits_to_double(lbits)
		sign = -1.0 if raw_h < 0 or (raw_h == 0.0 and raw_l < 0) else 1.0
		acc = DoubleDouble(abs(raw_h), abs(raw_l))
		factor = bits_to_double(fbits)
		for name, fn in scalers:
			y = sign * fn(acc, factor)
			if double_bits(y) != obits:
				counts[name] += 1
	for name in counts:
		print(f"{name} mismatches: {counts[name]}")

if __name__ == "__main__":
	run()
